### PR TITLE
[ntp]: Build 4.2.6 locally.

### PR DESCRIPTION
**Pull request by praveen-li**
Need to build NTP 4.2.6 locally due to a Bug in NTP 4.2.6.

Changes:
1.) Remove 'apt-get ntp' step.  File: build_debian.sh.
2.) Add NTP package as part of base image. File: slave.mk and in files/build_templates/sonic_debian_extension.j2.
3.) NTP Makefile for Sonic Build System. File: rules/ntp.mk.
4.) NTP Source Makefile.  File: src/ntp/Makefile.
5.) Patch to fix the issue.  File: src/ntp/patch.

Signed-off-by: Praveen Chaudhary<pchaudhary@linkedin.com>

**- What I did**
------------------------------------------------------------------------
I build NTP 4.2.6 locally by fetching dsc file from debian site. This is done to address a bug in NTP 4.2.6.

Bug Description: 
In NTP 1:4.2.6 code. File: ntpd/ntp_io.c. Function: process_routing_msgs(). 
Asyncio reader is disabled, if any error is returned from read call. Function read() may return ENOBUFS if kernel returns ENOBUFS in netlink due to memory constraints or due to link flap. Due to this ntp stops syncing time even when link comes back. Workaround was to restart NTP process, which is not acceptable. Kindle see below code:

    cnt = read(reader->fd, buffer, sizeof(buffer));
    
    if (cnt < 0) { 
          msyslog(LOG_ERR,
            "i/o error on routing socket %m - disabling");<<<<<<<<<<
        remove_asyncio_reader(reader);
        delete_asyncio_reader(reader);
    }
        return;
    }

To solve this, below fix is applied. Note: Fix exists in NTP 4.2.8, but NTP 4.2.8 is not available for Jessie. Also due to oher dependecies, it was decided to build NTP 4.2.6 with fix, instead of NTP 4.2.8

Code after Fix:
```4430     cnt = read(reader->fd, buffer, sizeof(buffer));
4431 
4432     if (cnt < 0) {
4433        if (errno == ENOBUFS) {
4434         msyslog(LOG_ERR, "routing socket reports: %m");
4435        } else {
4436         msyslog(LOG_ERR,
4437             "i/o error on routing socket %m - disabling");
4438         remove_asyncio_reader(reader);
4439         delete_asyncio_reader(reader);
4440        }
4441         return;
4442     }
```
 

**-
 How I did it**
------------------------------------------------------------------------
I build NTP package using below changes and applied point fix.

Changes:
1.) Remove 'apt-get ntp' step. File: build_debian.sh.
2.) Add NTP package as part of base image. File: slave.mk and in files/build_templates/sonic_debian_extension.j2.
3.) NTP Makefile for Sonic Build System. File: rules/ntp.mk.
4.) NTP Source Makefile. File: src/ntp/Makefile.
5.) Patch to fix the issue.   File: src/ntp/patch.

**- How to verify it**
------------------------------------------------------------------------
Ran Link Flap Operation:
```
Below Error was seen before fix:
syslog.93.gz:Feb  2 12:16:18.229161 falco-test-dut01 ERR ntpd[1222]: i/o error on routing socket No buffer space available - disabling
syslog.93.gz:Feb  2 12:20:45.429109 falco-test-dut01 ERR ntpd[27281]: i/o error on routing socket No buffer space available - disabling
```
Also command "ntpq -p" shows:
```# ntpq -p
No association ID's returned
```
After Fix:
We see new Error MSG:
```
syslog.9.gz:Feb  5 03:03:47.180823 falco-test-dut01 ERR ntpd[6153]: routing socket reports: No buffer space available
syslog.9.gz:Feb  5 03:03:47.240752 falco-test-dut01 ERR ntpd[6153]: routing socket reports: No buffer space available
syslog.9.gz:Feb  5 03:03:48.805217 falco-test-dut01 ERR ntpd[6153]: routing socket reports: No buffer space available
```
After fix "ntpq -p" output:
```admin@falco-test-dut01:/var/log$ ntpq -p
     remote           refid      st t when poll reach   delay   offset  jitter
==============================================================================
 10.0.0.1        .XFAC.          16 u    - 1024    0    0.000    0.000   0.000
 10.0.0.2        .XFAC.          16 u    - 1024    0    0.000    0.000   0.000
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: sonic-slave/Dockerfile [hunk 1]
```diff
@@ -237,6 +237,12 @@ RUN apt-get update && apt-get install -y \
         qemu-kvm \
         libvirt-bin
 
+# For ntp
+        autogen \
+        libopts25-dev \
+        pps-tools \
+        dh-apparmor
```
Comment (lguohan) [inline sonic-slave/Dockerfile]: ntp is installed base image and it should be built in stretch docker.

File: rules/ntp.mk [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+NTP_VERSION_SUFFIX = .p5+dfsg
+NTP_VERSION_SUFFIX_NUM = 7+deb8u2
+NTP_VERSION_FULL = $(NTP_VERSION_MAJOR)$(NTP_VERSION_SUFFIX)-$(NTP_VERSION_SUFFIX_NUM)
+NTP_VERSION = $(NTP_VERSION_MAJOR)$(NTP_VERSION_SUFFIX)
+
+export NTP_VERSION NTP_VERSION_FULL
+
+NTP = ntp_$(NTP_VERSION_FULL)_amd64.deb
+$(NTP)_SRC_PATH = $(SRC_PATH)/ntp
+SONIC_MAKE_DEBS += $(NTP)
```
Comment (lguohan) [inline rules/ntp.mk]: add SONIC_STRETCH_DEBS += $(NTP)

**Comment by lguohan** [comment]
please resolve conflict and rebase to latest master.

File: rules/ntp.mk (hunk 2 above)
Comment (praveen-li) [reply rules/ntp.mk]: Done.

**Comment by lguohan** [comment]
since stretch is already using 4.2.8, I am not sure why we are moving back to 4.2.6.

**Comment by praveen-li** [comment]
> since stretch is already using 4.2.8, I am not sure why we are moving back to 4.2.6.

Yup now after stretch changes, we do not need to build ntp 4.2.6 and have the point fix.
NTP 4.2.8 already have the fix for the bug which I was trying to address. We can discard this PR now.

**Comment by lguohan** [comment]
agree, can you open pr on 201803 branch? Looks like we need it there?

**Comment by praveen-li** [comment]
Done. Thanks,


Regards
Praveen

**Pull request closed without merging**